### DISCUSSION TITLE
Python: Retrieve (or log) Token Counts During Function Execution

### DISCUSSION THREAD

**Discussion by Gal-Gilor** [microsoft/semantic-kernel#5633] [Ideas]
Hi,

I’m trying to get a precise token counts from kernel, and planners, especially the Stepwise planner, during function execution. 

Unfortunately, I did not find the documentation to help with that. 

That said , I noticed that I can use PromptTemplateConfig, PromptTemplate, PromptTemplateEngine, SemanticFunctionConfig, and the context object to render the prompt. 

If it’s not yet feasible, is it possible to add to the ‘context’ another variable with the rendered prompt or the entire execution token counts (input, outputs both?).

Thoughts?

Thanks!

**Comment by moonbox3** [comment]
Hi @Gal-Gilor, which version of the SK Python package are you using? By the sound of the classes you're referring to, you're not using the latest beta package. I would highly recommend you upgrade to the latest package so you can take advantage of our latest enhancements. We have introduced a lot of changes to bring us closer to parity with dotnet. For example, functions now return a `FunctionResult` object that does contain the token count (both input and completion tokens) for OpenAI models.

The `usage` [key](https://github.com/microsoft/semantic-kernel/blob/f0b1d0df783d0420904a7e07c4957c5ce7d00c48/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L257C14-L257C19) is present in the `FunctionResult`'s metadata. As we move towards v1 we do have work items to improve/update our docs. This will be documented soon.

**Reply by nehalakhani** [reply]
Hi @moonbox3  - Is there alternative way to calculate token based on latest Semantic Kernel version 1.8.3

**Comment by Gal-Gilor** [comment]
Thanks for the quick response! Unfortunately, I am forced to used the older version. That said, I will bring that up.

**Reply by moonbox3** [reply]
Could you help us understand why you're forced to use the older version? Is it an internal requirement? We won't be able to provide any support for SK Python previous to our beta package. Please keep this in mind. Thanks.

**Reply by Gal-Gilor** [reply]
It is an internal requirement. You can consider this as resolved.
